File: src/main/java/minic/uilocal/visual/MiniCAstTreeModelFactory.java
package minic.uilocal;

import minic.uiapi.UiAstNodeVisualDto;
import minic.uiapi.UiStageVisualDto;

import java.util.ArrayList;
import java.util.List;

/**
 * 根据 AST visual DTO 生成树形行模型。
 */
public final class MiniCAstTreeModelFactory {
    /**
     * 创建 AST 树行模型。
     *
     * @param visual 当前阶段 visual DTO
     * @return 树行
     */
    public List<MiniCAstTreeLine> create(UiStageVisualDto visual) {
        if (visual.astRoot() == null) {
            return List.of();
        }
        ArrayList<MiniCAstTreeLine> lines = new ArrayList<>();
        append(visual.astRoot(), 0, lines);
        return List.copyOf(lines);
    }

    private void append(UiAstNodeVisualDto node, int depth, ArrayList<MiniCAstTreeLine> lines) {
        lines.add(new MiniCAstTreeLine(node.label(), depth, node.active(), node.range()));
        node.children().forEach(child -> append(child, depth + 1, lines));
    }
}

File: uiweb/src/visual/MiniCAstTreeModelFactory.ts
import type { JavaMirrorFile, UiAstNodeVisualDto, UiStageVisualDto } from "../translation/javaMirror";
import { MiniCAstTreeLine } from "./MiniCAstTreeLine";

export const miniCAstTreeModelFactoryMirror = {
  javaPath: "src/main/java/minic/uilocal/visual/MiniCAstTreeModelFactory.java",
  webPath: "uiweb/src/visual/MiniCAstTreeModelFactory.ts",
  packageName: "minic.uilocal.visual",
  exportName: "MiniCAstTreeModelFactory",
  kind: "class",
  imports: ["minic.uiapi.UiAstNodeVisualDto", "minic.uiapi.UiStageVisualDto", "java.util.List"],
  fields: [],
  methods: [{ name: "create", signature: "create(UiStageVisualDto visual)" }],
} as const satisfies JavaMirrorFile;

export class MiniCAstTreeModelFactory {
  static readonly mirror = miniCAstTreeModelFactoryMirror;

  create(visual: UiStageVisualDto | null | undefined): readonly MiniCAstTreeLine[] {
    if (!visual?.astRoot) {
      return [];
    }
    const lines: MiniCAstTreeLine[] = [];
    this.append(visual.astRoot, 0, lines);
    return lines;
  }

  private append(node: UiAstNodeVisualDto, depth: number, lines: MiniCAstTreeLine[]): void {
    lines.push(new MiniCAstTreeLine(node.label, depth, node.active, node.range));
    for (const child of node.children) {
      this.append(child, depth + 1, lines);
    }
  }
}

export default MiniCAstTreeModelFactory;
